pkg/resource/cidr_collection/hooks.go

```diff
@@ -0,0 +1,185 @@
+}
+
+// oldLocations returns a slice of CidrLocation pointer objects
+// to delete.
+func (rm *resourceManager) oldLocations(
+	latest *resource,
+) []*svcsdk.CidrCollectionChange {
+	if latest == nil {
+		return nil
+	}
+
+	locations := latest.ko.Spec.Locations
+	var oldLocations []*svcsdk.CidrCollectionChange
+	// Add removed locations to Changes using the DELETE_IF_EXISTS action
+	if len(locations) > 0 {
+		action := "DELETE_IF_EXISTS"
+		for _, rr := range locations {
+			change := &svcsdk.CidrCollectionChange{}
+			change.SetLocationName(*rr.LocationName)
+			change.SetCidrList(rr.CIDRList)
+			change.SetAction(action)
+			oldLocations = append(oldLocations, change)
+		}
+	}
+	return oldLocations
+}
+
+// newLocations returns a slice of CidrLocation pointer objects
+// with values set by the resource's corresponding spec field.
+func (rm *resourceManager) newLocations(
+	desired *resource,
+) ([]*svcsdk.CidrCollectionChange, error) {
+	if desired == nil {
+		return nil, nil
+	}
+
+	locations := desired.ko.Spec.Locations
+
+	var newLocations []*svcsdk.CidrCollectionChange
+	// Add locations to Changes using the PUT action
+	if len(locations) > 0 {
+		action := "PUT"
+		for _, rr := range locations {
+			if rr.LocationName == nil {
+				return nil, errors.New("InvalidInput: locationName is required in cidrLocation")
+			}
+			if rr.CIDRList == nil {
+				return nil, errors.New("InvalidInput: cidrList is required in cidrLocation")
+			}
+			change := &svcsdk.CidrCollectionChange{}
+			change.SetLocationName(*rr.LocationName)
+			change.SetCidrList(rr.CIDRList)
+			change.SetAction(action)
+			newLocations = append(newLocations, change)
+		}
+	}
+
+	return newLocations, nil
+}
+
+// customUpdateRecordSet is the custom implementation for
+// RecordSet resource's update operation.
+func (rm *resourceManager) customUpdateCidrCollection(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (updated *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateCidrCollection")
+	defer func() {
+		exit(err)
+	}()
+
+	if delta != nil {
+		// Do not proceed with update if an immutable field was updated
+		if immutableFieldChanges := rm.getImmutableFieldChanges(delta); len(immutableFieldChanges) > 0 {
+			msg := fmt.Sprintf("Immutable Spec fields have been modified: %s", strings.Join(immutableFieldChanges, ","))
+			return nil, ackerr.NewTerminalError(fmt.Errorf(msg))
+		}
+	}
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	input := &svcsdk.ChangeCidrCollectionInput{}
+	input.SetId(*desired.ko.Status.Collection.ID)
+	collectionVersion := *desired.ko.Status.Collection.Version
+	input.SetCollectionVersion(collectionVersion)
+
+	newLocations, err := rm.newLocations(desired)
+	if err != nil {
+		return nil, err
+	}
+	oldLocations := rm.oldLocations(latest)
```

Comment:
What if the desired state didn't change? are still remove and adding CIDR blocks at every reconciliation loop?

Reply:
I added a check to do nothing if nothing changes